File: app/src/main/java/jp/sblo/pandora/aGrep/SearchViewModel.java
package jp.sblo.pandora.aGrep;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

/**
 * ViewModel for managing search operations and UI state.
 */
public class SearchViewModel extends ViewModel {
    private final MutableLiveData<SearchUiState> uiState;
    private GrepEngine grepEngine;

    public SearchViewModel() {
        this.uiState = new MutableLiveData<SearchUiState>(SearchUiState.idle());
    }

    public LiveData<SearchUiState> getUiState() {
        return uiState;
    }

    /**
     * Initialize the grep engine with a Context.
     * Should be called from Activity/Fragment after ViewModel creation.
     */
    public void initializeEngine(Context context) {
        if (grepEngine == null) {
            grepEngine = new ExecutorGrepEngine(context);
        }
    }

    /**
     * Start a new search operation.
     */
    public void search(SearchRequest request) {
        if (grepEngine == null) {
            uiState.setValue(SearchUiState.error("Engine not initialized"));
            return;
        }

        uiState.setValue(SearchUiState.searching(new SearchProgress(
            request.query,
            0,
            0,
            null
        )));

        grepEngine.search(request, new GrepEngine.ProgressCallback() {
            @Override
            public void onProgress(SearchProgress progress) {
                uiState.postValue(SearchUiState.searching(progress));
            }

            @Override
            public void onComplete(SearchSummary summary) {
                uiState.postValue(SearchUiState.completed(summary));
            }

            @Override
            public void onError(String error) {
                uiState.postValue(SearchUiState.error(error));
            }
        });
    }

    /**
     * Cancel the current search operation.
     */
    public void cancelSearch() {
        if (grepEngine != null) {
            grepEngine.cancel();
        }
        uiState.setValue(SearchUiState.idle());
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (grepEngine != null) {
            grepEngine.shutdown();
        }
    }
}

File: src/jp/sblo/pandora/aGrep/SearchViewModel.kt
package jp.sblo.pandora.aGrep

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.asLiveData
import java.util.ArrayList
import java.util.Collections
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class SearchViewModel(
    private val grepEngine: CoroutineGrepEngine
) : ViewModel() {

    private val job = SupervisorJob()
    private val scope = CoroutineScope(job + Dispatchers.Default)
    private var searchJob: Job? = null

    private val _state = MutableStateFlow(SearchUiState())
    val state: StateFlow<SearchUiState> = _state
    val stateLiveData: LiveData<SearchUiState> = state.asLiveData()

    fun startSearch(request: SearchRequest) {
        val current = _state.value
        if (current.isSearching && current.query == request.query) {
            return
        }
        searchJob?.cancel()
        _state.value = SearchUiState(query = request.query, isSearching = true)
        searchJob = scope.launch {
            try {
                val summary = grepEngine.search(request) { progress ->
                    _state.update { previous ->
                        val combined = if (progress.newMatches.isEmpty()) {
                            previous.results
                        } else {
                            val newList = ArrayList(previous.results.size + progress.newMatches.size)
                            newList.addAll(previous.results)
                            newList.addAll(progress.newMatches)
                            newList
                        }
                        previous.copy(
                            filesProcessed = progress.filesProcessed,
                            matchesFound = progress.matchesFound,
                            statusMessage = progressStatus(progress.filesProcessed, progress.matchesFound),
                            results = combined
                        )
                    }
                }
                val sorted = ArrayList(summary.results)
                Collections.sort(sorted, GrepView.Data())
                _state.update {
                    it.copy(
                        isSearching = false,
                        isCompleted = true,
                        statusMessage = progressStatus(summary.filesProcessed, summary.matchesFound),
                        results = sorted
                    )
                }
            } catch (cancelled: CancellationException) {
                _state.update {
                    it.copy(isSearching = false, isCancelled = true)
                }
            } catch (throwable: Throwable) {
                _state.update {
                    it.copy(isSearching = false, errorMessage = throwable.message)
                }
            }
        }
    }

    fun cancelSearch() {
        searchJob?.cancel()
    }

    override fun onCleared() {
        super.onCleared()
        job.cancel()
    }

    private fun progressStatus(files: Int, matches: Int): String =
        "$files files • $matches matches"

    data class SearchUiState(
        val query: String = "",
        val isSearching: Boolean = false,
        val isCompleted: Boolean = false,
        val isCancelled: Boolean = false,
        val filesProcessed: Int = 0,
        val matchesFound: Int = 0,
        val results: List<GrepView.Data> = emptyList(),
        val statusMessage: String? = null,
        val errorMessage: String? = null
    )

    class Factory(
        private val grepEngine: CoroutineGrepEngine
    ) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            if (modelClass.isAssignableFrom(SearchViewModel::class.java)) {
                @Suppress("UNCHECKED_CAST")
                return SearchViewModel(grepEngine) as T
            }
            throw IllegalArgumentException("Unknown ViewModel class: " + modelClass.name)
        }
    }
}
